ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/StateTransition.java

```diff
@@ -66,9 +73,18 @@ public BeaconState initiate(
       BeaconStateWithCache state, BeaconBlock block, boolean validate_state_root)
       throws StateTransitionException {
     try {
+      final UnsignedLong previousSlot = state.getSlot();
       // Process slots (including those with no blocks) since block
       process_slots(state, block.getSlot(), printEnabled);
 
+      epochMetrics.ifPresent(
+          metrics -> {
+            final UnsignedLong endEpoch = compute_epoch_of_slot(block.getSlot());
+            if (endEpoch.compareTo(compute_epoch_of_slot(previousSlot)) > 0) {
+              metrics.onEpoch(state);
```

Comment:
Why not just call metrics.onEpoch in process_slots, right after process_epoch(), or in process_epoch() right at the end? We basically do the same check to run process_epoch anyway

Reply:
That went off into static land so means passing more stuff through, but I've learnt more since making that decisions so gone back and rethought.

In the latest version I've changed `StateTransition.process_slots` to not be static so it can access `epochMetrics` and call `onMetrics` itself.  I think it's quite nice but is a change to the pattern of using static methods for spec functions.